x-pack/plugin/wildcard/src/main/java/org/elasticsearch/xpack/wildcard/mapper/WildcardFieldMapper.java

```diff
@@ -216,187 +229,175 @@ public WildcardFieldType clone() {
+            Automaton ngramAutomaton = ngramRegex.toAutomaton(maxDeterminizedStates);
 
-
+            return createApproxAndVerifyQuery(ngramAutomaton, automaton, value);
+        }
+        
+        private static String toLowerCase(String string) {
+            return lowercaseNormalizer.normalize(null, string).utf8ToString();
         }
+        
+        
+        Query createApproxAndVerifyQuery(Automaton simplifiedAutomaton, Automaton verifyAutomaton, String expression) {
 
+            Query ngramApproximation = createApproximationQueryFromAutomaton(simplifiedAutomaton);
 
-        @Override
-        public Query wildcardQuery(String wildcardPattern, RewriteMethod method, QueryShardContext context) {
-            PatternStructure patternStructure = new PatternStructure(wildcardPattern);
-            ArrayList<String> tokens = new ArrayList<>();
-
-            for (int i = 0; i < patternStructure.fragments.length; i++) {
-                String fragment = patternStructure.fragments[i];
-                int fLength = fragment.length();
-                if (fLength == 0) {
-                    continue;
+            AutomatonQueryOnBinaryDv verifyingQuery = new AutomatonQueryOnBinaryDv(name(), expression, verifyAutomaton);
+            if (ngramApproximation != null && !(ngramApproximation instanceof MatchAllDocsQuery)) {
+                // We can accelerate execution with the ngram query
+                BooleanQuery.Builder verifyingBuilder = new BooleanQuery.Builder();
+                verifyingBuilder.add(new BooleanClause(ngramApproximation, Occur.MUST));
+                verifyingBuilder.add(new BooleanClause(verifyingQuery, Occur.MUST));
+                return verifyingBuilder.build();
+            }
+            return verifyingQuery;
+        }
+        
+
+        private Query createApproximationQueryFromAutomaton(Automaton simplifiedAutomaton) {
+            FiniteStringsIterator iterator = new FiniteStringsIterator(simplifiedAutomaton);
+            BooleanQuery.Builder result = new BooleanQuery.Builder();
+            int numAlternativePathsConsidered = 0;
+            for (IntsRef finiteString; (finiteString = iterator.next()) != null;) {
+                // Efficient simplification is tough - we need to create a query that
+                // covers ALL of the possible regex permutations.
+                // However, FiniteStringsIterator is a depth-first iteration
+                // so EVERY alternative path would have to be considered before we could
+                // determine a common section all the paths shared in order to simplify.
+                // So, we simplify up-front by:
+                // 1) replacing all repetitions eg (foo)* with a single invalid char in the regex string used to build the automaton
+                // 2) lowercasing all concrete values so searches like [Ee][Nn][Cc][Oo][Dd][Ee][Dd] don't fork myriad paths
```

Comment:
I'd like to see the solution without lowercasing the ngram index. We can try to optimize lowercased search in a follow up but for now we should ensure that we handle all types of automaton without relying on normalization ?

Reply:
I think I need to understand better how an FSI approach can be made to work with an approach that tries to find articulation points in the automaton. The former relies on FSI to chase down all the possible paths while the latter takes direct control of exploring the graph paths to understand the branching (like Nik's code). 
If we try to make life simple by using FSI only then some up-front case normalisation is useful to avoid searches like the `[Pp][Oo][Ww][Ee][Rr][Ss][Hh][Ee][Ll][Ll]` example from [this blog](https://socprime.com/en/blog/elastic-for-security-analysts-part-1-searching-strings/) blowing the BooleanQuery limits.